src/Orleans/GrainDirectory/IGrainDirectory.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans.GrainDirectory
+{
+
+    /// <summary>
+    /// Recursive distributed operations on grain directories.
+    /// Each operation may forward the request to a remote owner, increasing the hopCount.
+    /// 
+    /// The methods here can be called remotely (where extended by IRemoteGrainDirectory) or
+    /// locally (where extended by ILocalGrainDirectory)
+    /// </summary>
+    interface IGrainDirectory
+    {
+        /// <summary>
+        /// Record a new grain activation by adding it to the directory.
+        /// <para>This method must be called from a scheduler thread.</para>
+        /// </summary>
+        /// <param name="address">The address of the new activation.</param>
+        /// <param name="singleActivation">If true, use single-activation mode: 
+        /// If there is already an activation registered for this grain, then the new activation will
+        /// not be registered and the address of the existing activation will be returned.
+        /// Otherwise, the passed-in address will be returned.</param>
+        /// <param name="hopCount">Counts recursion depth across silos</param>
+        /// <returns>The registered address and the version associated with this directory mapping.</returns>
+        Task<AddressAndTag> RegisterAsync(ActivationAddress address, bool singleActivation, int hopCount = 0);
```

Comment:
I think I prefer to keep this an explicit parameter, to make code at call sites easier to understand.

As far as I can tell, the only place that this is called with singleActivation:false is to register clients in the grain directory. We could pass an "isClient" parameter instead of singleActivation to reduce confusion with the other potential meanings of "single instance".